pyfr/integrators/dual/phys/steppers.py

```diff
@@ -1,25 +1,139 @@
 # -*- coding: utf-8 -*-
 
+import math
+
 from pyfr.integrators.dual.phys.base import BaseDualIntegrator
 
 
 class BaseDualStepper(BaseDualIntegrator):
-    pass
 
+    def step(self, t, dt):
+        scoeffs = [self.stepper_coeffs(s, dt)
+                   for s in range(self.nstages)]
+
+        for s, (sc, tc) in enumerate(zip(scoeffs, self.c)):
+            self.pseudointegrator._init_stage(s, sc)
+            self.pseudointegrator.pseudo_advance(t + dt*tc)
+
+        self._finalize_step()
+
+    def _finalize_step(self):
+        pass
+
+
+class BaseBDFStepper(BaseDualStepper):
+    nstages = 1
+    stage_nregs = 0
+    c = [1]
+
+    @property
+    def stepper_nregs(self):
+        return len(self.stepper_static_coeffs) - 1
+
+    def _finalize_step(self):
+        psnregs = self.pseudointegrator.pseudo_stepper_nregs
+        snregs = self.pseudointegrator.stepper_nregs
 
-class DualBDF2Stepper(BaseDualStepper):
+        # Rotate the source registers to the right by one
+        self.pseudointegrator._regidx[psnregs:psnregs + snregs] = (
+            self.pseudointegrator._stepper_regidx[-1:]
+            + self.pseudointegrator._stepper_regidx[:-1]
+        )
+
+        # Copy the current soln into the first source register
+        self.pseudointegrator._add(
+            0, self.pseudointegrator._stepper_regidx[0],
+            1, self.pseudointegrator._idxcurr
+        )
+
+    def stepper_coeffs(self, s, dt):
+        return [1] + [sc/dt for sc in self.stepper_static_coeffs]
+
+
+class DualBDF2Stepper(BaseBDFStepper):
     stepper_name = 'bdf2'
     stepper_order = 2
-    stepper_coeffs = [-1.5, 2.0, -0.5]
+    stepper_static_coeffs = [-1.5, 2.0, -0.5]
 
 
-class DualBDF3Stepper(BaseDualStepper):
+class DualBDF3Stepper(BaseBDFStepper):
     stepper_name = 'bdf3'
     stepper_order = 3
-    stepper_coeffs = [-11.0/6.0, 3.0, -1.5, 1.0/3.0]
+    stepper_static_coeffs = [-11.0/6.0, 3.0, -1.5, 1.0/3.0]
+
+
+class BaseDIRKStepper(BaseDualStepper):
+    stepper_nregs = 1
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if self.fsal:
+            self.b = self.a[-1]
 
+        self.c = [sum(row) for row in self.a]
 
-class DualBackwardEulerStepper(BaseDualStepper):
+    @property
+    def stage_nregs(self):
+        return self.nstages
+
+    def stepper_coeffs(self, s, dt):
+        return [self.a[s][s], -1/dt, 1/dt] + self.a[s][:s]
+
+    def _finalize_step(self):
+        if not self.fsal:
+            self._obtain_solution()
+
+        # Copy the new solution into stepper register
+        self.pseudointegrator._add(
+            0, self.pseudointegrator._stepper_regidx[0],
```

Comment:
We should try and restructure this so it does not depend on private members.

Reply:
What is your suggestion?

Reply:
Open to suggestions, but we really should not be prodding around with the pseudo integrators registers.

Reply:
When we cancel bdf steppers I can move this into pseudo class and call from phys level. I can also move now but then we'll have two/three functions that are similar.